cmd/ingester/app/consumer/consumer.go

```diff
@@ -44,133 +46,246 @@ type Consumer struct {
+	// reference consumerReady and make sure to close it, because later consumerReady may reference a new object
+	ready := c.consumerReady
+	defer ready.close()
+
+	for {
+		select {
+		case <-ctx.Done():
+			c.logger.Error("ctx canceled")
+			return
+		default:
+			c.logger.Info("Topic", zap.Strings("topic", topic))
+			if err := c.internalConsumer.Consume(ctx, topic, c); err != nil {
+				c.logger.Error("Error from consumer", zap.Error(err))
+			}
+
+			c.consumerReady.close()
+
+			c.consumerReady = &consumerReady{
+				readyCh: make(chan struct{}, 1),
+			}
+
+			// check if context was cancelled, signaling that the consumer should stop
+			if ctx.Err() != nil {
+				if c.cancel != nil {
+					c.cancel()
+				}
+			}
+		}
+	}
+}
+
+// markDeadlock mark consume is deadlocked
+func (c *Consumer) markDeadlock() {
+	c.deadlockNotifyOnce.Do(func() {
+		close(c.deadlockNotifyCh)
+	})
+}
+
+// Deadlock return a consume deadlock notify chan
+func (c *Consumer) Deadlock() <-chan struct{} {
+	return c.deadlockNotifyCh
 }
 
 // Close closes the Consumer and underlying sarama consumer
 func (c *Consumer) Close() error {
 	// Close the internal consumer, which will close each partition consumers' message and error channels.
 	c.logger.Info("Closing parent consumer")
 	err := c.internalConsumer.Close()
-
-	c.logger.Debug("Closing deadlock detector")
 	c.deadlockDetector.close()
 
+	if c.cancel != nil {
+		c.cancel()
+	}
+
 	c.logger.Debug("Waiting for messages and errors to be handled")
 	c.doneWg.Wait()
 
 	return err
 }
 
-// handleMessages handles incoming Kafka messages on a channel
-func (c *Consumer) handleMessages(pc sc.PartitionConsumer) {
-	c.logger.Info("Starting message handler", zap.Int32("partition", pc.Partition()))
-	c.partitionMapLock.Lock()
-	c.partitionsHeld++
-	c.partitionsHeldGauge.Update(c.partitionsHeld)
-	c.partitionMapLock.Unlock()
-	defer func() {
-		c.closePartition(pc)
-		c.partitionMapLock.Lock()
-		c.partitionsHeld--
-		c.partitionsHeldGauge.Update(c.partitionsHeld)
-		c.partitionMapLock.Unlock()
-		c.doneWg.Done()
-	}()
+// Setup is run at the beginning of a new session, before ConsumeClaim
+func (c *Consumer) Setup(sarama.ConsumerGroupSession) error {
+	// Mark the consumer as ready
+	c.consumerReady.markReady()
+	return nil
+}
 
-	msgMetrics := c.newMsgMetrics(pc.Partition())
+// Cleanup is run at the end of a session, once all ConsumeClaim goroutines have exited
+func (c *Consumer) Cleanup(sarama.ConsumerGroupSession) error {
+	return nil
+}
+
+// ConsumeClaim must start a consumer loop of ConsumerGroupClaim's Messages().
+func (c *Consumer) ConsumeClaim(session sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
+	c.partitionMetrics(claim.Partition()).startCounter.Inc(1)
+
+	return c.startMessagesLoop(session, claim)
+}
+
+func (c *Consumer) startMessagesLoop(session sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
+	return c.handleMessages(session, claim)
+}
```

Comment:
I think the indirection is not really necessary. If we want to improve readability on what `handleMessages` does, then a short comment should suffice.

```suggestion
	return c.handleMessages(session, claim)
}
```

Reply:
Thanks, the code has been modified